From b12b4c28b4c4a76cbc906b703ee5a694a082ab74 Mon Sep 17 00:00:00 2001
From: Christian Hopps <chopps@labn.net>
Date: Tue, 8 Apr 2025 05:15:53 +0000
Subject: [PATCH] mgmtd: remove bogus "hedge" code which corrupted active
 candidate DS

Say you have 2 mgmtd frontend sessions (2 vtysh's) the first one is long
running and is actively changing the global candidate datastore (DS),
the second one starts and exits, this code would then copy running
back over the candidate, blowing away any changes made by the first
session.

(the long running session could technically be any user)

Instead we need to trust the various cleanup code that already exits.
For example in the commit_cfg_reply on success candidate is copied to
running, and on failure *for implicit commit* running is copied back to
candidate clearing the change. This leaves the non-implicit
configuration changes in this case we actually want candidate to keep
it's changes in transactional cases, in the other case of pending commit
during a file read the code restores candidate (if needed) on exit from
"config terminal", with this call stack:

 vty_config_node_exit()
   nb_cli_pending_commit_check()
     nb_cli_classic_commit()
       nb_candidate_commit_prepare() [fail] -> copy running -> candidate
       nb_candidate_commit_apply() -> copy candidate -> running

fixes #18541

Signed-off-by: Christian Hopps <chopps@labn.net>
---
 mgmtd/mgmt_fe_adapter.c | 6 ------
 1 file changed, 6 deletions(-)

diff --git a/mgmtd/mgmt_fe_adapter.c b/mgmtd/mgmt_fe_adapter.c
index 8d59198803..d077e08679 100644
--- mgmt_fe_adapter.c	2025-09-28 17:45:35.958313082 +0000
+++ mgmt_diff.c	2025-09-28 18:04:40.184270480 +0000
@@ -115,16 +115,7 @@
 
 static void
 mgmt_fe_session_cfg_txn_cleanup(struct mgmt_fe_session_ctx *session)
-{
-	/*
-	 * Ensure any uncommitted changes in Candidate DS
-	 * is discarded.
-	 */
-	if (session->config_count > 0) {
-		mgmt_ds_copy_dss(mm->running_ds, mm->candidate_ds, false);
-		session->config_count = 0;
-	}
-
+{ 
 	/*
 	 * Destroy the actual transaction created earlier.
 	 */